internal/job/api.go

```diff
@@ -25,7 +25,7 @@ func (e *Executor) startJobAPI() (cleanup func(), err error) {
 		return cleanup, fmt.Errorf("creating job API socket path: %v", err)
 	}
 
-	srv, token, err := jobapi.NewServer(e.shell.Logger, socketPath, e.shell.Env)
+	srv, token, err := jobapi.NewServer(e.shell.Logger, socketPath, e.shell.Env, e.redactors)
```

Comment:
without getting too borrow-checker-rewrite-it-in-rust-y, what's the lifetime of `e.redactors` as used by the job api? if something adds a new redactor mid-job, what happens? is `replacer.Mux` goroutine safe?

i _think_ because all slices are references, if a redactor is added it should be fine, but if someone calls `e.redactors = ...` then it'll break? is that true?

Reply:
Yep, this code does seem dangerous. Redactors used to be created in each hook, but now they are created for the life of the job executor.

> but if someone calls `e.redactors = ...` then it'll break? is that true?

The redactors will still work in that the job logs will still redact the set of values that the redactors had prior to the reassignment. The garbage collector won't clean them up either, as there are still references to the redactors in
1. `e.shell.Writer`
1. `e.shell.Logger.WriterLogger`

In any case, the `Mux` will still be referred to in the job api server.

What will break is that a new hook won't be able to add the value of new environment variables that match the pattern to the redactor.

So as long as the job is still writing to `e.shell.Writer` and `e.shell.Logger`, previous redaction would still work.

Ultimately, this danger also exists for `e.shell.Env`. If someone write `e.shell = ...` to a new reference, then a lot of things will break. I think this is an indication that the redactors should be inside the shell, though. I think I will make a change to do that.

Reply:
yeah no that's totally fine - the risk is largely theoretical as is, happy for this to be done later.